File: src/Frontend/monoton_client/lib/Models/base/Message.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:monoton_client/Models/Dto/Response/ApplicationUser.dart';
import 'package:monoton_client/Models/base/Chat.dart';

part 'Message.g.dart';



@JsonSerializable()
class Message {
  @JsonKey(name: 'id')
  late String id;

  @JsonKey(name: 'content')
  late List<int> content;

  @JsonKey(name: 'sender')
  late ApplicationUser sender;

  @JsonKey(name: 'targetChat')
  late Chat targetChat;

  @JsonKey(name: 'createdAt', toJson: _dateTimeToJson, fromJson: _dateTimeFromJson)
  late DateTime createdAt;

  @JsonKey(name: 'messageType')
  late MessageType messageType;

  @JsonKey(name: 'resourceAddress')
  late String resourceAddress;

  @JsonKey(name: 'messageResourceType')
  late MessageResourceType messageResourceType;

  @JsonKey(name: 'isReplay')
  late bool isReplay;

  @JsonKey(name: 'metadata')
  late String metadata;

  @JsonKey(name: 'seenBy')
  late List<ApplicationUser> seenBy;

  Message();

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);

  Map<String, dynamic> toJson() => _$MessageToJson(this);

  static DateTime _dateTimeFromJson(String json) => DateTime.parse(json);

  static String _dateTimeToJson(DateTime dateTime) => dateTime.toIso8601String();
}

enum MessageType {
  @JsonValue('TEXT')
  TEXT,
  @JsonValue('IMAGEM')
  IMAGE,
  @JsonValue('VIDEO')
  VIDEO,
  @JsonValue('FILE')
  FILE,
  @JsonValue('VOICE')
  VOICE,
  @JsonValue('DELETED')
  DELETED,
  @JsonValue('NULL')
  NULL,
}

enum MessageResourceType {
  @JsonValue('INTERNAL')
  INTERNAL,
  @JsonValue('EXTERNAL')
  EXTERNAL,
}

File: src/Backend/Monotone/Models/base/Message.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Monotone.Models.Enums;

namespace Monotone.Models.@base;

public record Message
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public string Id { get; set; } = null!;
    public byte[] Content { get; set; } = Array.Empty<byte>();
    public ApplicationUser Sender { get; set; } = null!;
    public Chat TargetChat { get; set; } = null!;
    [DataType(DataType.Date)]
    public DateTime CreatedAt { get; set; }
    public MessageType MessageType { get; set; }
    public string ResourceAddress { get; set; } = string.Empty;
    
    public MessageResourceType MessageResourceType { get; set; } 
    public bool IsReplay { get; set; }
    public string Metadata { get; set; } = string.Empty;
    public List<ApplicationUser> SeenBy { get; set; } = new();
}